### PR TITLE
Fix CallToolResult handling across all SDKs

### LINKED ISSUE #937: SDK does not accept CallToolResult type of MCP/SDK type for Tool Response
The GitHub Copilot SDK needs to support the MCP SDK `CallToolResult` type for tool responses.

**Team**: Office Product Group
**Linked Release**: https://github.com/github/releases/issues/7690

### PR DESCRIPTION
**Pull request by stephentoub**
## Problem

When a tool handler returns an MCP `CallToolResult` object (`{ content: [{type, text/data/resource}], isError?: boolean }`), all four SDKs JSON-serialize it and send the raw JSON string over RPC. The LLM then sees `'{"content":[{"type":"text","text":"actual result"}]}'` instead of the actual tool output.

Fixes #937

## Approach

Each SDK's tool result normalization path now detects the `CallToolResult` shape **before** the JSON-serialize fallback and converts it to the SDK's native `ToolResultObject` format:

- **Text** content blocks → `textResultForLlm` (multiple blocks joined with `\n`)
- **Image** content blocks → `binaryResultsForLlm` with `type: "image"`
- **Resource** content blocks → text goes to `textResultForLlm`, blob goes to `binaryResultsForLlm`
- `isError: true` → `resultType: "failure"`

The .NET SDK additionally handles **Microsoft.Extensions.AI** content types (`TextContent`, `DataContent`, and unknown `AIContent` subtypes via `AIJsonUtilities` serialization), since the MCP C# SDK's `McpClientTool` returns these types from `InvokeCoreAsync`.

## Changes by SDK

**Node.js** (`nodejs/src/types.ts`, `session.ts`, `client.ts`)
- `isCallToolResult()` type guard + `convertCallToolResult()` converter
- Integrated into both `_executeToolAndRespond` (session API) and `normalizeToolResultV2` (client API)
- Defensive guards for malformed input (optional chaining on resource, typeof check on text)

**Python** (`python/copilot/tools.py`)
- `_is_call_tool_result()` + `_convert_call_tool_result()` (private helpers)
- Integrated into `_normalize_result`

**Go** (`go/definetool.go`)
- `convertCallToolResult()` with JSON marshal/unmarshal fallback for typed Go structs
- Integrated into `normalizeResult`

**.NET** (`dotnet/src/Types.cs`, `Session.cs`, `Client.cs`)
- `ToolResultObject.TryConvertFromCallToolResult(object?)` — detects `JsonElement` with CallToolResult shape
- `ToolResultObject.TryConvertFromAIContent(object?)` — handles `TextContent`, `DataContent`, `IEnumerable<AIContent>`, with fallback serialization for unknown subtypes
- Integrated into both `ExecuteToolAndRespondAsync` (session) and client v2 tool response path

## Tests

- **Node.js**: 21 unit tests in `nodejs/test/call-tool-result.test.ts`
- **Go**: 9 subtests in `TestConvertCallToolResult`
- **Python**: 13 tests in `TestCallToolResult` class
- Coverage includes: text-only, multiple text, isError mapping, image content, resource text/blob, mixed content, empty content, malformed input rejection, and edge cases

### CODE REVIEW

File: go/definetool.go [hunk 1]
```diff
@@ -99,7 +106,101 @@ func normalizeResult(result any) (ToolResult, error) {
...
+		TextResultForLLM: strings.Join(textParts, "\n"),
+		ResultType:       resultType,
+	}
+	if len(binaryResults) > 0 {
+		tr.BinaryResultsForLLM = binaryResults
+	}
+	return tr, true
+}
+
+// generateSchemaForTypegenerates a JSON schema map from a Go type using reflection.
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply go/definetool.go]: Fixed — added the missing space in the comment.

File: python/copilot/tools.py [hunk 2]
```diff
@@ -274,3 +279,51 @@ def default(obj: Any) -> Any:
...
+                        data=resource["blob"],
+                        mime_type=resource.get("mimeType", "application/octet-stream"),
+                        type="resource",
+                        description=resource.get("uri", ""),
+                    )
+                )
+
+    return ToolResult(
+        text_result_for_llm="\n".join(text_parts),
+        result_type="failure" if call_result.get("isError") else "success",
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply python/copilot/tools.py]: Fixed — changed to `is True` for consistency with the other SDKs which all strictly check for boolean true.

File: python/copilot/tools.py [hunk 3]
```diff
@@ -274,3 +279,51 @@ def default(obj: Any) -> Any:
...
+                    mime_type=block.get("mimeType", ""),
+                    type="image",
+                )
+            )
+        elif block_type == "resource":
+            resource = block.get("resource", {})
+            if resource.get("text"):
+                text_parts.append(resource["text"])
+            if resource.get("blob"):
+                binary_results.append(
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply python/copilot/tools.py:312]: Fixed — added isinstance(resource, dict) guard as part of the broader type-safety improvements in Comment 1.

File: python/copilot/tools.py [hunk 4]
```diff
@@ -274,3 +279,51 @@ def default(obj: Any) -> Any:
...
+                    ToolBinaryResult(
+                        data=resource["blob"],
+                        mime_type=resource.get("mimeType", "application/octet-stream"),
+                        type="resource",
+                        description=resource.get("uri", ""),
+                    )
+                )
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply python/copilot/tools.py]: Fixed — added isinstance checks for all field accesses in _convert_call_tool_result to guard against non-string values, consistent with the other SDKs' defensive handling.

File: nodejs/src/index.ts [hunk 5]
```diff
@@ -10,7 +10,13 @@
 
 export { CopilotClient } from "./client.js";
 export { CopilotSession, type AssistantMessageEvent } from "./session.js";
-export { defineTool, approveAll, SYSTEM_PROMPT_SECTIONS } from "./types.js";
+export {
+    defineTool,
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply nodejs/src/index.ts:14]: Already addressed — Prettier was run on this file in a previous commit and CI formatting checks pass.

File: python/e2e/test_tools_unit.py [hunk 6]
```diff
@@ -6,7 +6,12 @@
 from pydantic import BaseModel, Field
 
 from copilot import define_tool
-from copilot.tools import ToolInvocation, ToolResult, _normalize_result
+from copilot.tools import (
+    ToolInvocation,
+    ToolResult,
+    _normalize_result,
+    _is_call_tool_result,
+)
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply python/e2e/test_tools_unit.py:14]: Already addressed — ruff format was run on this file in a previous commit and CI formatting checks pass.

File: go/definetool.go [hunk 7]
```diff
@@ -99,7 +106,101 @@ func normalizeResult(result any) (ToolResult, error) {
...
+
+	contentRaw, exists := m["content"]
+	if !exists {
+		return ToolResult{}, false
+	}
+
+	contentSlice, ok := contentRaw.([]any)
+	if !ok {
+		return ToolResult{}, false
+	}
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply go/definetool.go:129]: The function already has a JSON marshal/unmarshal fallback (lines 113-122) for typed structs. When the value is already a map[string]any, the content slice will always be []any from JSON unmarshaling. The current handling is sufficient.

**Review by stephentoub — COMMENTED** [review]
Fixed the inconsistency.

**Comment by SteveSandersonMS** [comment]
If I understand correctly, the reason why people might want conversion of `CallToolResult` from custom tools is that they are implementing their own MCP host and exposing the MCP tools as custom tools, e.g.,

```ts
defineTool("inner_tool", {
     handler: async (params) => {
         return await someMcpClient.callTool("inner_tool", params);
     },
 });
```

Is that right?

If that's the case, then what would you think about tweaking this slightly so that we expose the `CallToolResult` converter as a utility that's optional but still very easy to use? I think it would be a relatively small change, leading to APIs something like:

```ts
// TypeScript
handler: async ({ query }) => {
     const mcpResult = await mcpClient.callTool("inner_tool", { query });
     return convertCallToolResult(mcpResult);
 }
```

```cs
 // C#
 async (string query) => {
     var mcpResult = await mcpClient.CallToolAsync("inner_tool", new { query });
     return ToolResultObject.FromCallToolResult(mcpResult);
 }
```

```py
 # Python
 @define_tool("my_tool", description="Delegates to an MCP server")
 async def my_tool(params: MyParams, invocation: ToolInvocation) -> ToolResult:
     mcp_result = await mcp_client.call_tool("inner_tool", {"query": params.query})
     return convert_call_tool_result(mcp_result)
```

```go
 // Go
 func(params MyParams, inv copilot.ToolInvocation) (copilot.ToolResult, error) {
     mcpResult, err := mcpClient.CallTool(ctx, "inner_tool", params)
     if err != nil {
         return copilot.ToolResult{}, err
     }
     return copilot.ConvertCallToolResult(mcpResult)
 }
```

I'm wondering because using custom tools to wrap MCP client calls is a valid scenario but not universal, and having duck-typed auto-conversion could lead to surprises if people's custom tool outputs happen to be a similar shape, e.g.:

```ts
defineTool("list_components", {
     description: "Lists UI components in a directory",
     parameters: z.object({ dir: z.string() }),
     handler: async ({ dir }) => {
         return {
             content: [
                 { type: "button", label: "Submit", file: "submit.tsx" },
                 { type: "input", label: "Search", file: "search.tsx" },
             ],
         };
     },
 });
```

Hopefully having conversion as a utility wouldn't make things meaningfully harder for people passing through MCP client calls, and clarifies the separation between custom tools and MCP.

What do you think? Am I understanding the scenario?

**Comment by stephentoub** [comment]
@SteveSandersonMS, thanks for the feedback. I've addressed it by making the conversions explicit (except for .NET, where M.E.AI that we already reference includes a type we can special-case).

**Review by SteveSandersonMS — APPROVED** [review]
Great!
